### DISCUSSION TITLE
RNDIS + lwIP + FreeRTOS on CH32V307 not working

### DISCUSSION THREAD

**Discussion by LeeChunHei** [hathach/tinyusb#1891] [Q&A]
Hi,

I'm working on enabling rndis/ecm on CH32V307, however, it didn't work.
I tried to build a CDC+MSC example with this hardware setup and it's working, so I expect the hardware and the porting to be good.

When I set the debug level as 2 and run the program, I saw these logs.

`
SystemClk:144000000
FreeRTOS Kernel Version:V10.4.6
USBD init on controller 0
sizeof(usbd_device_t) = 83
sizeof(tu_fifo_t) = 20
sizeof(tu_edpt_stream_t) = 104
NET init
USBD Bus Reset : High Speed

USBD Setup Received 80 06 00 01 00 00 40 00 
  Get Descriptor Device
  Queue EP 80 with 18 bytes ...
USBD Xfer Complete on EP 80 with 18 bytes
  Queue EP 00 with 0 bytes ...
USBD Xfer Complete on EP 00 with 0 bytes
USBD Bus Reset : High Speed

USBD Setup Received 00 05 01 00 00 00 00 00 
  Set Address
USBD Xfer Complete on EP 80 with 0 bytes

USBD Setup Received 80 06 00 01 00 00 12 00 
  Get Descriptor Device
  Queue EP 80 with 18 bytes ...
USBD Xfer Complete on EP 80 with 18 bytes
  Queue EP 00 with 0 bytes ...
USBD Xfer Complete on EP 00 with 0 bytes

USBD Setup Received 80 06 00 0F 00 00 05 00 
  Get Descriptor BOS
  Stall EP0

USBD Setup Received 80 06 00 0F 00 00 05 00 
  Get Descriptor BOS
  Stall EP0

USBD Setup Received 80 06 00 0F 00 00 05 00 
  Get Descriptor BOS
  Stall EP0

USBD Setup Received 80 06 00 02 00 00 09 00 
  Get Descriptor Configuration[0]
  Queue EP 80 with 9 bytes ...

USBD Setup Received 80 06 00 02 00 00 09 00 
  Get Descriptor Configuration[0]
  Queue EP 80 with 9 bytes ...

USBD Setup Received 80 06 00 02 00 00 09 00 
  Get Descriptor Configuration[0]
  Queue EP 80 with 9 bytes ...
USBD Bus Reset : High Speed

USBD Setup Received 80 06 00 01 00 00 40 00 
  Get Descriptor Device
  Queue EP 80 with 18 bytes ...
USBD Xfer Complete on EP 80 with 9 bytes
  Queue EP 00 with 0 bytes ...

USBD Setup Received 80 06 00 01 00 00 40 00 
  Get Descriptor Device
  Queue EP 80 with 18 bytes ...
USBD Xfer Complete on EP 80 with 18 bytes
  Queue EP 00 with 0 bytes ...
USBD Bus Reset : High Speed

USBD Setup Received 00 05 0A 00 00 00 00 00 
  Set Address

USBD Setup Received 00 05 0B 00 00 00 00 00 
  Set Address
USBD Xfer Complete on EP 80 with 0 bytes

USBD Setup Received 80 06 00 01 00 00 12 00 
  Get Descriptor Device
  Queue EP 80 with 18 bytes ...
USBD Xfer Complete on EP 80 with 18 bytes
  Queue EP 00 with 0 bytes ...
USBD Xfer Complete on EP 00 with 0 bytes

USBD Setup Received 80 06 00 0F 00 00 05 00 
  Get Descriptor BOS
  Stall EP0

USBD Setup Received 80 06 00 0F 00 00 05 00 
  Get Descriptor BOS
  Stall EP0

USBD Setup Received 80 06 00 0F 00 00 05 00 
  Get Descriptor BOS
  Stall EP0

USBD Setup Received 80 06 00 02 00 00 09 00 
  Get Descriptor Configuration[0]
  Queue EP 80 with 9 bytes ...

USBD Setup Received 80 06 00 02 00 00 09 00 
  Get Descriptor Configuration[0]
  Queue EP 80 with 9 bytes ...

USBD Setup Received 80 06 00 02 00 00 09 00 
  Get Descriptor Configur
`

Anyone had the idea on how to solve this?

**Comment by scaprile** [comment]
How do you setup RNDIS ? What do you enable ? What do you disable ? Where do you try ? What is code ? What are the system logs ? Did you put a USB sniffer ? How did you configure lwiP ? How did you configure FreeRTOS ? Did you try your FreeRTOS+lwIP configuration with an Ethernet micro to be sure it is working ? Do you have a DHCP server ? With RNDIS you need to assign an IP address to the USB host, unless you use locally administered addresses (unfortunately I don't remember how to do that)

**Reply by LeeChunHei** [reply]
The code I use is in this [repo](https://github.com/LeeChunHei/tinyusb_ch32v307_test), I had successfully tried the CDC+MSC example before, so this shouldn't be hardware or tinyusb porting problem.

> How do you setup RNDIS ? 

I just use the [usb_descriptor.c](https://github.com/hathach/tinyusb/blob/master/examples/device/net_lwip_webserver/src/usb_descriptors.c) and [tusb_config.h](https://github.com/hathach/tinyusb/blob/master/examples/device/net_lwip_webserver/src/tusb_config.h) in the tinyusb example folder

> What do you enable ? 

`CFG_TUSB_MCU => OPT_MCU_CH32V307`
`CFG_TUSB_OS => OPT_OS_FREERTOS`
`CFG_TUD_ENABLED => 1`
`CFG_TUD_ECM_RNDIS => 1`

> What do you disable ? 

`CFG_TUD_NCM => (1-CFG_TUD_ECM_RNDIS)`

> Where do you try ?

On CH32V307 

> What is code ? 

[here](https://github.com/LeeChunHei/tinyusb_ch32v307_test)

>What are the system logs ?

code 10 
![image](https://user-images.githubusercontent.com/24974375/218377182-39e309e5-0d55-412f-a6db-bacd96240030.png)
 
> Did you put a USB sniffer ? 

I use wireshark and the dump is [here](https://github.com/LeeChunHei/tinyusb_ch32v307_test/blob/master/usb_capture/rndis.pcapng)

> How did you configure lwiP ?

I use this config [lwipopts.h](https://github.com/LeeChunHei/tinyusb_ch32v307_test/blob/master/lwip/port/lwipopts.h) and [usb_ethernetif.c](https://github.com/LeeChunHei/tinyusb_ch32v307_test/blob/master/lwip/port/usb_ethernetif.c) to connect lwip and tinyusb
 
> How did you configure FreeRTOS ? 

With this [https://github.com/LeeChunHei/tinyusb_ch32v307_test/blob/master/User/FreeRTOSConfig.h](FreeRTOSConfig.h)

> Did you try your FreeRTOS+lwIP configuration with an Ethernet micro to be sure it is working ? 

I didn't try that.
But I don't think it's the problem with FreeRTOS+lwIP, 
since `tud_network_recv_cb` never triggers.
So nothing notifies lwIP to do anything.

> Do you have a DHCP server ? 

Yes, using the DHCP server code provided in tinyusb lib/network folder

> With RNDIS you need to assign an IP address to the USB host, unless you use locally administered addresses (unfortunately I don't remember how to do that)

The windows account I'm working on has administrator privelage, and this is also not working under Linux.

**Reply by scaprile** [reply]
>>     Where do you try ?
> On CH32V307

I doubt you try the device on itself, sorry for a non-precise question. Later on you say you tried on Windows and Linux

The device has to assign an address to the host, you don't need admin privileges in your computers. Please post your Linux log: `dmesg -w` when you plug your device

**Reply by LeeChunHei** [reply]
dmesg -w log as follow:
[27710.672729] usb 3-8: new high-speed USB device number 2 using xhci_hcd
[27711.121372] usb 3-8: New USB device found, idVendor=cafe, idProduct=4020, bcdDevice= 1.01
[27711.121376] usb 3-8: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[27711.121377] usb 3-8: Product: TinyUSB Device
[27711.121378] usb 3-8: Manufacturer: TinyUSB
[27711.121378] usb 3-8: SerialNumber: 123456     
[27711.968739] rndis_host 3-8:1.0: RNDIS_MSG_QUERY(0x01010101) failed, -110
[27711.968743] rndis_host 3-8:1.0: rndis get ethaddr, -110
[27711.996410] rndis_host: probe of 3-8:1.0 failed with error -110
looks like it failed to query something?

**Reply by scaprile** [reply]
Make sure you have properly configured TinyUSB for FreeRTOS, and setup proper MAC addresses; sniff the networking interchange, follow next answer below. I'm too lazy to traverse your repo in search of what you might have possibly left out or done wrong.

**Comment by scaprile** [comment]
Please compare your setup with this working example: https://github.com/cesanta/mongoose/tree/7.9/examples/stm32/nucleo-f429zi-usb-rndis

**Comment by LeeChunHei** [comment]
I think I can narrow down this problem to the porting with CH32V307.
I tried the same setup on my ‎i.MX RT1011 board and windows can recognize the rndis device without any problem.

**Comment by hathach** [comment]
ch32v307 is a new port, and still considered as WIP. I have no surprise if an advanced/complicated class driver such as rndis/ecm doesn't work on it yet. I don't have time to at ch32v307 for now, though **have you tried the stock net_lwip_webserver example without freeRTOS**, just to see if that work ? it could be the dcd race condition with preempted RTOS.

**Reply by LeeChunHei** [reply]
I tried CH32V307 without FreeRTOS and still not working

**Reply by LeeChunHei** [reply]
This is quite interesting, cause I failed to make rndis device work on CherryUSB also.
Not sure if this is mcu design problem or just bug in the porting.

**Reply by hathach** [reply]
It is probably only ch32 driver (software) issue, there is no much of hw difference between msc and usbnet. usbnet may have more complicated usage and probably needs more works e.g with intense payload etc.. I haven't tested much with this port and unfortunately couldn't have time to help. Please keep posting anything you find interesting with the ports. Maybe others would help.

**Reply by Krasutski** [reply]
Dear @hathach, I reproduced this issue on [ch32f20x](https://github.com/hathach/tinyusb/pull/2278) MCU. Also, I observed not stable work of ch32v307/dcd_usbhs.c for other examples on ch32f205 MCU. I can take a look and try to fix DCD for this platform if this task isn't in progress by somebody. 
If the task is still open, please let me know

**Reply by Krasutski** [reply]
I am a bit fixed and cleanup DCD for CH32 High Speed, now it works more stable. RNDIS example works other dual examples like msc+cdc, dual CDC, and MSC dual LUN work a long time without ASSERTs
DCD driver is common for all CH32 High Speed series it should work for all MCUs. But I tested it only for CH32F2x. It would be good to double-check it on CH32V3/V2. 
PR: #2392
